segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeperstorage/LogStorage.java

```diff
@@ -0,0 +1,203 @@
+            // If there is no ledger, create a new one.
+            LedgerData data = manager.createLedgerAt(this.name, (int) offset);
+            this.dataMap.put((int) offset, data);
+            return data;
+        }
+    }
+
+    /**
+     * Add a new BK ledger and metadata at a given offset.
+     * @param offset The starting offset represented by the ledger.
+     * @param ledgerData metadata of the ledger.
+     */
+    synchronized void addToList(int offset, LedgerData ledgerData) {
+
+        // If we are replacing an existing ledger, adjust the length
+
+        LedgerData older = this.dataMap.put(offset, ledgerData);
+        if (older != null) {
+            this.length -= older.getLedgerHandle().getLength();
+        }
+        this.length += ledgerData.getLedgerHandle().getLength();
+    }
+
+    /**
+     * Increase length of the LogStorage as a side effect of the write operation.
+     * This is just a cache operation. The length is not persisted.
+     *
+     * @param size size of data written.
+     */
+    synchronized void increaseLengthBy(int size) {
+        this.length += size;
+    }
+
+    synchronized CompletableFuture<Void> deleteAllLedgers() {
+        return CompletableFuture.allOf(
+                this.dataMap.entrySet().stream().map(entry -> manager.deleteLedger(entry.getValue().getLedgerHandle())).toArray(CompletableFuture[]::new));
+    }
+
+    LedgerData getLastLedgerData() {
+        if (this.dataMap.isEmpty()) {
+            return null;
+        } else {
+            return this.dataMap.entrySet().stream().max(Comparator.comparingInt(Map.Entry::getKey)).get().getValue();
+        }
+    }
+
+    synchronized void markSealed() {
+        this.sealed = true;
+    }
+
+    /**
+     * Creates a list of curator transaction for merging source LogStorage in to this.
+     * @param source Name of the source ledger.
+     * @return list of curator operations.
+     */
+    synchronized List<CuratorOp> addLedgerDataFrom(LogStorage source) {
+        List<CuratorOp> retVal = source.dataMap.entrySet().stream().map(entry -> {
+            int newKey = (int) (entry.getKey() + this.length);
+            this.dataMap.put(newKey, entry.getValue());
+            try {
+                return manager.createAddOp(this.name, newKey, entry.getValue());
+            } catch (StreamSegmentException e) {
+                //TODO: throw proper exceptions.
+                return null;
+            }
+        }).collect(Collectors.toList());
+
+        this.length += source.length;
+        return retVal;
+    }
+
+    synchronized LedgerData getLedgerDataForReadAt(long offset) {
+        if (offset >= length) {
+            throw new CompletionException( new BadOffsetException(this.getName(), length, offset));
+        }
+        Optional<Map.Entry<Integer, LedgerData>> found = dataMap.entrySet().stream().filter(entry -> (entry.getKey() <= offset) && (offset < (entry.getKey() + entry.getValue().getLength()))).findFirst();
```

Comment:
Here an in deleteLastLedgerEntry it is using an ordering based on keys. Perhaps dataMap should be a ConcurrentSkipListMap instead.

Reply:
Replaced with `ConcurrentSkipListMap `